pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -663,23 +662,23 @@ void messageReceived(MessageIdData messageId, ByteBuf headersAndPayload, ClientC
             // thread while the message processing happens
             listenerExecutor.execute(() -> {
                 for (int i = 0; i < numMessages; i++) {
-                    Message msg;
-                    try {
-                        msg = internalReceive();
-                    } catch (PulsarClientException e) {
-                        log.warn("[{}] [{}] Failed to dequeue the message for listener", topic, subscription, e);
-                        return;
-                    }
-
-                    try {
-                        if (log.isDebugEnabled()) {
-                            log.debug("[{}][{}] Calling message listener for message {}", topic, subscription, msg);
+                    internalReceiveAsync().handle((msg, ex) -> {
```

Comment:
[messageReceived](https://github.com/yahoo/pulsar/blob/master/pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java#L631) adds message into ```incomingMessages``` and then ```listener-thread``` calls ```internalReceiveAsync()``` which either ```poll(0, TimeUnit.MILLISECONDS)``` or adds call to ```pendingReceives``` incase ```incomingMessages``` is cleared by connectionOpened()/redeliverUnackedMsgs() (so, thread will not be blocked in case msg is not present into queue).
So, i think it should not pass the same message always. right?